File: src/rect.lua
local Rect = {}
Rect.__index = Rect

function Rect.new(x, y, width, height)
	local self = setmetatable({}, Rect)

	self.x = x
	self.y = y
	self.width = width
	self.height = height

	return self
end

function Rect:intersects(rect)

	local tw = this.width
	local th = this.height
	local rw = rect.width
	local rh = rect.height
	if (rw <= 0 or rh <= 0 or tw <= 0 or th <= 0) then
		return false
	end

	local tx = this.x
	local ty = this.y
	local rx = rect.x
	local ry = rect.y
	rw = rw + rx
	rh = rh + ry
	tw = tw + tx
	th = th + ty

	return ((rw < rx or rw > tx) and
		(rh < ry or rh > ty) and
		(tw < tx or tw > rx) and
		(th < ty or th > ry))
end

return Rect

File: src/rect.js
define(function () {
	function Rect(x, y, width,height) {
		this.x = +x || 0;
		this.y = +y || 0;
		this.width = +width || 1;
		this.height = +height || 1;
	}

	Rect.prototype.fill = function (ctx) {
		ctx.fillRect(Math.floor(this.x), Math.floor(this.y), this.width, this.height);
	};

	Rect.prototype.stroke = function (ctx) {
		ctx.strokeRect(Math.floor(this.x), Math.floor(this.y), this.width, this.height);
	};

	Rect.prototype.contains = function (x, y) {
		return (
			(x >= this.x && x <= this.x + this.width) &&
			(y >= this.y && y <= this.y + this.height)
		);
	};

	/* modified from the java.awt.Rect source code */
	Rect.prototype.intersects = function (rect) {
		var tw, th, rw, rh, tx, ty, rx, ry;

		tw = this.width;
		th = this.height;
		rw = rect.width;
		rh = rect.height;
		if (rw <= 0 || rh <= 0 || tw <= 0 || th <= 0) {
			return false;
		}
		tx = this.x;
		ty = this.y;
		rx = rect.x;
		ry = rect.y;
		rw += rx;
		rh += ry;
		tw += tx;
		th += ty;
		//      overflow || intersect
		return ((rw < rx || rw > tx) &&
			(rh < ry || rh > ty) &&
			(tw < tx || tw > rx) &&
			(th < ty || th > ry));
	};

	return Rect;
});
